products-shared-kernel/src/main/scala/dev/atedeg/mdm/products/Utils.scala

```diff
@@ -1,26 +1,60 @@
 package dev.atedeg.mdm.products
 
+import scala.compiletime.*
+
 import cats.data.NonEmptyList
-import dev.atedeg.mdm.utils.{PositiveNumber, coerce}
+import cats.kernel.Order
 import eu.timepit.refined.predicates.all.Positive
 
-import scala.compiletime.*
+import dev.atedeg.mdm.utils.{ coerce, PositiveNumber }
 
 type OneOf[T <: Tuple] = T match
   case (t *: EmptyTuple) => t
   case (t *: ts) => t | OneOf[ts]
 
+@SuppressWarnings(Array("scalafix:DisableSyntax.asInstanceOf"))
 inline def all[T <: Tuple]: NonEmptyList[OneOf[T]] = inline erasedValue[T] match
-  case _:(n *: EmptyTuple) =>
+  case _: (n *: EmptyTuple) =>
     val v = checkInt[n](constValue[n])
-    NonEmptyList.one(toGrams(v)).asInstanceOf[NonEmptyList[OneOf[T]]]
-  case _:(n *: gs) =>
+    NonEmptyList.one(v).asInstanceOf[NonEmptyList[OneOf[T]]]
+  case _: (n *: gs) =>
     val v = checkInt[n](constValue[n])
-    NonEmptyList(toGrams(v), all[gs].toList).asInstanceOf[NonEmptyList[OneOf[T]]]
+    NonEmptyList(v, all[gs].toList).asInstanceOf[NonEmptyList[OneOf[T]]]
   case _ => compiletime.error("Cannot work on a tuple with elements that are not Grams")
 
-private inline def checkInt[T](inline n: T): Int = inline erasedValue[T] match
+@SuppressWarnings(Array("scalafix:DisableSyntax.asInstanceOf"))
+inline private def checkInt[T](inline n: T): Int = inline erasedValue[T] match
   case _: Int => n.asInstanceOf[Int]
   case _ => compiletime.error(codeOf(n) + " is not an int")
 
-private[products] def toGrams(n: Int): Grams = Grams(coerce[Int, Positive](n))
\ No newline at end of file
+private[products] def toGrams(n: Int): Grams = Grams(coerce[Int, Positive](n))
+
+given Order[Grams] with
+  def compare(x: Grams, y: Grams): Int = Order[Int].compare(x.n.value, y.n.value)
+
+extension (cheeseType: CheeseType)
+
+  /**
+   * Creates a [[Product product]] with the given [[CheeseType cheese type]] if it can find a
+   * weight amongst the allowed ones that matches the given predicate.
+   */
+  def withWeight(predicate: Int => Boolean): Option[Product] =
+    // Somehow Scala does not understand that a function (A => B) could also be treated as a function
+    // [a <: A] => a => B and won't compile unless I explicitly convert the function myself...
+    def p = [t <: Int] => (n: t) => predicate(n)
```

Comment:
Following the scala coding convention, the generics types must start with capital letter
```suggestion
    def p = [T <: Int] => (n: T) => predicate(n)
```